inference-engine/src/mkldnn_plugin/nodes/mkldnn_matrix_nms_node.cpp

```diff
@@ -0,0 +1,405 @@
+    });
+
+    InferenceEngine::parallel_for(m_numBatches, [&](size_t batchIdx) {
+        size_t batchOffset = batchIdx * m_realNumClasses * m_realNumBoxes;
+        BoxInfo* batchFilteredBox = m_filteredBoxes.data() + batchOffset;
+        auto& numPerClass = m_numPerBatchClass[batchIdx];
+        auto numDet = std::accumulate(numPerClass.begin(), numPerClass.end(), 0);
+        auto start_offset = numPerClass[0];
+
+        for (size_t i = 1; i < numPerClass.size(); i++) {
+            auto offset_class = m_classOffset[i];
+            for (size_t j = 0; j < numPerClass[i]; j++) {
+                batchFilteredBox[start_offset + j] = batchFilteredBox[offset_class + j];
+            }
+            start_offset += numPerClass[i];
+        }
+        auto keepNum = numDet;
+        if (m_keepTopk > -1) {
+            auto k = static_cast<size_t>(m_keepTopk);
+            if (keepNum > k)
+                keepNum = k;
+        }
+
+        std::partial_sort(batchFilteredBox, batchFilteredBox + keepNum, batchFilteredBox + numDet, [](const BoxInfo& lhs, const BoxInfo rhs) {
+            return lhs.score > rhs.score || (lhs.score == rhs.score && lhs.classIndex < rhs.classIndex) ||
+                   (lhs.score == rhs.score && lhs.classIndex == rhs.classIndex && lhs.index < rhs.index);
+        });
+        m_numPerBatch[batchIdx] = keepNum;
+    });
+
+    auto startOffset = m_numPerBatch[0];
+    for (size_t i = 1; i < m_numPerBatch.size(); i++) {
+        auto offset_batch = i * m_realNumClasses * m_realNumBoxes;
+        for (size_t j = 0; j < m_numPerBatch[i]; j++) {
+            m_filteredBoxes[startOffset + j] = m_filteredBoxes[offset_batch + j];
+        }
+        startOffset += m_numPerBatch[i];
+    }
+
+    m_filteredBoxes.resize(startOffset);
+
+    if (m_sortResultAcrossBatch) { /* sort across batch */
+        if (m_sortResultType == MatrixNmsSortResultType::SCORE) {
+            parallel_sort(m_filteredBoxes.begin(), m_filteredBoxes.end(), [](const BoxInfo& l, const BoxInfo& r) {
+                return (l.score > r.score) || (l.score == r.score && l.batchIndex < r.batchIndex) ||
+                       (l.score == r.score && l.batchIndex == r.batchIndex && l.classIndex < r.classIndex) ||
+                       (l.score == r.score && l.batchIndex == r.batchIndex && l.classIndex == r.classIndex && l.index < r.index);
+            });
+        } else if (m_sortResultType == MatrixNmsSortResultType::CLASSID) {
+            parallel_sort(m_filteredBoxes.begin(), m_filteredBoxes.end(), [](const BoxInfo& l, const BoxInfo& r) {
+                return (l.classIndex < r.classIndex) || (l.classIndex == r.classIndex && l.batchIndex < r.batchIndex) ||
+                       (l.classIndex == r.classIndex && l.batchIndex == r.batchIndex && l.score > r.score) ||
+                       (l.classIndex == r.classIndex && l.batchIndex == r.batchIndex && l.score == r.score && l.index < r.index);
+            });
+        }
+    }
```

Comment:
just question
why we shouldn't apply sorting if ```m_sortResultAcrossBatch == false``` like in MultiClassNms?
https://github.com/openvinotoolkit/openvino/blob/c594d0787ebe07153e29e84b3d36724700739b91/inference-engine/src/mkldnn_plugin/nodes/mkldnn_multiclass_nms.cpp#L238-L245

Reply:
The **Matrix NMS** already keeps a sorted results in each batch during previous computation.
https://github.com/openvinotoolkit/openvino/blob/c594d0787ebe07153e29e84b3d36724700739b91/inference-engine/src/mkldnn_plugin/nodes/mkldnn_matrix_nms_node.cpp#L318-L343